flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -1488,8 +1493,34 @@ private void internalOfferSlotsToJobManager(JobTable.Connection jobManagerConnec
             JobID jobId,
             JobMasterGateway jobMasterGateway,
             JobMasterId jobMasterId,
-            Collection<SlotOffer> offeredSlots) {
+            Collection<SlotOffer> offeredSlots,
+            int offerId) {
         return (Iterable<SlotOffer> acceptedSlots, Throwable throwable) -> {
+            // check if this is the latest offer
+            if (offerId != offerCounter) {
```

Comment:
I think a global slot offer counter does not work. We need per job counter because otherwise it is not guaranteed that we will get a slot offer response for a given sub set of slots (for a given job).